File: packages/ik_rules/test/bank_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:ik_content/ik_content.dart';
import 'package:ik_parity/ik_parity.dart';
import 'package:ik_rules/ik_rules.dart';
import 'package:test/test.dart';

import 'support/fixtures.dart';

LocationRow _loc(String id, [String? mapId]) {
  return LocationRow({'Location ID': id, 'Map ID': ?mapId});
}

Object? _loadRawDatabase() {
  final file = File('../../content/data/game-database.json');
  return jsonDecode(file.readAsStringSync());
}

void main() {
  test('Town Bank, Citadel Bank, and future *_bank locations have a bank', () {
    expect(locationHasBank(_loc('LOC-0034', 'MAP-0006')), isTrue);
    expect(locationHasBank(_loc('LOC-0035', 'MAP-0007')), isTrue);
    expect(locationHasBank(_loc('LOC-0002')), isFalse);
    expect(locationHasBank(_loc('LOC-0013')), isFalse);
    expect(locationHasBank(_loc('LOC-0027')), isFalse);
    expect(locationHasBank(_loc('LOC-0024', 'MAP-0006')), isFalse);
    expect(locationHasBank(_loc('LOC-0009', 'MAP-0001')), isFalse);
    expect(locationHasBank(null), isFalse);
    expect(
      locationHasBank(
        LocationRow({
          'Location ID': 'LOC-9999',
          'Internal Key': 'west_bank',
          'Display Name': 'West Bank',
        }),
      ),
      isTrue,
    );
  });

  test('prepared databases expose deposit-box thievery at every bank', () {
    final loaded = prepareDatabase(_loadRawDatabase());
    final depositActs = loaded.launch.activities.where((row) => row.poolId == depositBoxPoolId);
    final bankLocs = loaded.launch.locations.where(locationLooksLikeBank);
    for (final location in bankLocs) {
      expect(
        depositActs.any((row) => row.locationId == location.locationId),
        isTrue,
        reason: location.locationId,
      );
    }

    final withFuture = withBankDepositBoxActivities(
      GameDatabase({
        ...loaded.launch.raw,
        'Locations': [
          ...loaded.launch.locations.map((row) => row.raw),
          {
            'Location ID': 'LOC-9999',
            'Internal Key': 'west_bank',
            'Display Name': 'West Bank',
            'Map ID': 'MAP-0006',
            'Location Type': 'Settlement',
            'Status': 'Planned',
            'Release Phase': 'Launch',
          },
        ],
      }),
    );
    expect(
      withFuture.activities.any(
        (row) => row.poolId == depositBoxPoolId && row.locationId == 'LOC-9999',
      ),
      isTrue,
    );
  });

  test('deposits, withdraws, and refuses gold', () {
    final db = databaseOf(loadParityFixtures('save/roundtrip').first);
    var save = createNewSave(db, 0).copyWith(
      inventory: const [InventoryStack(itemId: 'ITEM-0002', quantity: 5)],
      bank: const <InventoryStack>[],
    );

    final deposited = depositToBank(save, 0, 3);
    expect(deposited.ok, isTrue);
    save = deposited.save!;
    expect(save.inventory.single.quantity, 2);
    expect(save.bank.single.quantity, 3);

    final gold = depositToBank(
      save.copyWith(inventory: const [InventoryStack(itemId: 'ITEM-0001', quantity: 10)]),
      0,
      10,
    );
    expect(gold.ok, isFalse);
    expect(gold.reason, 'Gold cannot be deposited.');

    final withdrawn = withdrawFromBank(save, 0, 2);
    expect(withdrawn.ok, isTrue);
    expect(withdrawn.save!.bank.single.quantity, 1);
    expect(withdrawn.save!.inventory.single.quantity, 4);
  });
}

File: src/game/inventory/bank.test.ts
import { describe, expect, it } from 'vitest'
import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { prepareDatabase } from '../data/loadDatabase'
import { createNewSave } from '../save/saveStore'
import { INVENTORY_SLOT_LIMIT } from './capacity'
import {
  depositToBank,
  locationHasBank,
  stackIsUnbankableGold,
  withdrawFromBank,
} from './bank'
import { GOLD_ITEM_ID } from './gold'

const rawDatabase = JSON.parse(
  readFileSync(resolve(process.cwd(), 'content/data/game-database.json'), 'utf8'),
)

describe('bank storage', () => {
  const { launch, launchIndexes } = prepareDatabase(rawDatabase)

  it('is at Town, Castle, and Citadel — including their districts', () => {
    const byId = launchIndexes.locationsById
    expect(locationHasBank(byId.get('LOC-0002'))).toBe(true)
    expect(locationHasBank(byId.get('LOC-0013'))).toBe(true)
    expect(locationHasBank(byId.get('LOC-0027'))).toBe(true)
    expect(locationHasBank(byId.get('LOC-0024'))).toBe(true)
    expect(locationHasBank(byId.get('LOC-0014'))).toBe(true)
    expect(locationHasBank(byId.get('LOC-0028'))).toBe(true)
    expect(locationHasBank(byId.get('LOC-0009'))).toBe(false)
    expect(locationHasBank(undefined)).toBe(false)
  })

  it('deposits and withdraws stacks, and refuses gold', () => {
    let save = {
      ...createNewSave(launch),
      inventory: [{ itemId: 'ITEM-0002', quantity: 5 }],
      bank: [] as { itemId: string; quantity: number }[],
    }

    const deposited = depositToBank(save, 0, 3)
    expect(deposited.ok).toBe(true)
    if (!deposited.ok) return
    save = deposited.save
    expect(save.inventory).toEqual([{ itemId: 'ITEM-0002', quantity: 2 }])
    expect(save.bank).toEqual([{ itemId: 'ITEM-0002', quantity: 3 }])
    expect(save.gold).toBe(createNewSave(launch).gold)

    const gold = depositToBank(
      { ...save, inventory: [{ itemId: GOLD_ITEM_ID, quantity: 10 }] },
      0,
      10,
    )
    expect(gold.ok).toBe(false)
    if (gold.ok) return
    expect(gold.reason).toBe('Gold stays on you.')
    expect(stackIsUnbankableGold({ itemId: GOLD_ITEM_ID })).toBe(true)

    const withdrawn = withdrawFromBank(save, 0, 2)
    expect(withdrawn.ok).toBe(true)
    if (!withdrawn.ok) return
    expect(withdrawn.save.bank).toEqual([{ itemId: 'ITEM-0002', quantity: 1 }])
    expect(withdrawn.save.inventory).toEqual([{ itemId: 'ITEM-0002', quantity: 4 }])
  })

  it(`caps the chest at ${INVENTORY_SLOT_LIMIT} slots`, () => {
    const save = {
      ...createNewSave(launch),
      inventory: [{ itemId: 'ITEM-0002', quantity: 1 }],
      bank: Array.from({ length: INVENTORY_SLOT_LIMIT }, (_, index) => ({
        itemId: `ITEM-fake-${index}`,
        quantity: 1,
      })),
    }
    const blocked = depositToBank(save, 0, 1)
    expect(blocked.ok).toBe(false)
    if (blocked.ok) return
    expect(blocked.reason).toBe(`Bank is full (${INVENTORY_SLOT_LIMIT} slots).`)
  })
})
